{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9f297c69",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "uv 0.7.13\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: gdown [-h] [-V] [-O OUTPUT] [-q] [--fuzzy] [--id] [--proxy PROXY]\n",
      "             [--speed SPEED] [--no-cookies] [--no-check-certificate]\n",
      "             [--continue] [--folder] [--remaining-ok] [--format FORMAT]\n",
      "             [--user-agent USER_AGENT]\n",
      "             url_or_id\n",
      "\n",
      "positional arguments:\n",
      "  url_or_id             url or file/folder id (with --id) to download from\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -V, --version         display version (default: None)\n",
      "  -O OUTPUT, --output OUTPUT\n",
      "                        output file name/path; end with \"/\"to create a new\n",
      "                        directory (default: None)\n",
      "  -q, --quiet           suppress logging except errors (default: False)\n",
      "  --fuzzy               (file only) extract Google Drive's file ID (default:\n",
      "                        False)\n",
      "  --id                  [DEPRECATED] flag to specify file/folder id instead of\n",
      "                        url (default: False)\n",
      "  --proxy PROXY         <protocol://host:port> download using the specified\n",
      "                        proxy (default: None)\n",
      "  --speed SPEED         download speed limit in second (e.g., '10MB' ->\n",
      "                        10MB/s) (default: None)\n",
      "  --no-cookies          don't use cookies in ~/.cache/gdown/cookies.txt\n",
      "                        (default: False)\n",
      "  --no-check-certificate\n",
      "                        don't check the server's TLS certificate (default:\n",
      "                        False)\n",
      "  --continue, -c        resume getting partially-downloaded files while\n",
      "                        skipping fully downloaded ones (default: False)\n",
      "  --folder              download entire folder instead of a single file (max\n",
      "                        50 files per folder) (default: False)\n",
      "  --remaining-ok        (folder only) asserts that is ok to download max 50\n",
      "                        files per folder. (default: False)\n",
      "  --format FORMAT       Format of Google Docs, Spreadsheets and Slides.\n",
      "                        Default is Google Docs: 'docx', Spreadsheet: 'xlsx',\n",
      "                        Slides: 'pptx'. (default: None)\n",
      "  --user-agent USER_AGENT\n",
      "                        User-Agent to use for downloading file. (default:\n",
      "                        None)\n"
     ]
    }
   ],
   "source": [
    "uv --version\n",
    "gdown --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "134d9dc3",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "uv run gdown \"https://drive.google.com/uc?id=1Q0zg9zqTaZUvt42uxzE_0gcfnFvjwi33\" -O gnomix/pretrained_gnomix_models.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca8aed9a",
   "metadata": {},
   "source": [
    "Preprocess the GVCF into a format that is compatible with WhatsHap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd343c2",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/mismatch_removed/dup_removed/skipped:\t337580708/5583944/0/200294/0/0/0\n"
     ]
    }
   ],
   "source": [
    "bcftools norm --threads 24 -m -any -f reference/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta reference/EG100001.merged.g.vcf.gz | \\\n",
    "  bcftools view --threads 24 -v snps -m2 -M2 -Oz -o reference/EG100001.snps.vcf.gz\n",
    "\n",
    "bcftools index --threads 24 -t reference/EG100001.snps.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef26b0fc",
   "metadata": {},
   "source": [
    "Might need to filter the bam for only primary alignments\n",
    "\n",
    "Below filters for:\n",
    "- not primary alignment (0x100)\n",
    "- supplementary alignment (0x800)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6683804d",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "whatshap phase -o output/EG100001.phased.vcf.gz --reference=reference/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta reference/EG100001.snps.vcf.gz reference/EG100001.bam > WhatsHap.log 2>&1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ede64930",
   "metadata": {},
   "source": [
    "Liftover to hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e8e4ec4d",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to /tmp/bcftools.2IuQyX\n",
      "INFO/AS_InbreedingCoeff is handled by AGR rule\n",
      "INFO/AS_QD is handled by AGR rule\n",
      "INFO/MLEAC is handled by AGR rule\n",
      "INFO/MLEAF is handled by AGR rule\n",
      "FORMAT/GP is handled by AGR rule\n",
      "FORMAT/PG is handled by AGR rule\n",
      "FORMAT/PL is handled by AGR rule\n",
      "Lines   total/swapped/reference added/rejected:\t4450819/25773/772/208367\n",
      "Merging 2 temporary files\n",
      "Done\n",
      "Cleaning\n"
     ]
    }
   ],
   "source": [
    "docker run -it --rm -v \"$(pwd)\":/data -v /tmp:/tmp -w /data clarity001/bcftools-liftover:latest \\\n",
    "    /bin/bash -c \"bcftools +liftover -Ou output/EG100001.phased.vcf.gz -- -s reference/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta -f reference/hg19.fa -c reference/hg38ToHg19.over.chain.gz | bcftools sort -Oz -o output/EG100001.phased.hg19.vcf.gz -W=tbi\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c04d259",
   "metadata": {},
   "source": [
    "Test gnomix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fa8f96c7",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "...\n",
      "--------------------------------------------------------------------------------\n",
      "-----------------------------------  Gnomix  -----------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "When using this software, please cite: \n",
      "Helgi Hilmarsson, Arvind S Kumar, Richa Rastogi, Carlos D Bustamante, \n",
      "Daniel Mas Montserrat, Alexander G Ioannidis: \n",
      "\"High Resolution Ancestry Deconvolution for Next Generation Genomic Data\" \n",
      "https://www.biorxiv.org/content/10.1101/2021.09.19.460980v1\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "Error: Incorrect number of arguments.\n",
      "Usage when training a model from scratch:\n",
      "   $ python3 gnomix.py <query_file> <output_basename> <chr_nr> <phase> <genetic_map_file> <reference_file> <sample_map_file>\n",
      "Usage when using a pre-trained model:\n",
      "   $ python3 gnomix.py <query_file> <output_basename> <chr_nr> <phase> <path_to_model>\n"
     ]
    }
   ],
   "source": [
    "uv run gnomix/gnomix.py --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab54f264",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'gnomix/config.yaml' -> './config.yaml'\n"
     ]
    }
   ],
   "source": [
    "cp gnomix/config.yaml ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "104e057c",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "...\n",
      "--------------------------------------------------------------------------------\n",
      "-----------------------------------  Gnomix  -----------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "When using this software, please cite: \n",
      "Helgi Hilmarsson, Arvind S Kumar, Richa Rastogi, Carlos D Bustamante, \n",
      "Daniel Mas Montserrat, Alexander G Ioannidis: \n",
      "\"High Resolution Ancestry Deconvolution for Next Generation Genomic Data\" \n",
      "https://www.biorxiv.org/content/10.1101/2021.09.19.460980v1\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "Launching in pre-trained mode...\n",
      "Loading model...\n",
      "Launching inference...\n",
      "Loading and processing query file...\n",
      "- Number of SNPs from model: 317408\n",
      "- Number of SNPs from file: 62013\n",
      "- Number of intersecting SNPs: 46033\n",
      "- Percentage of model SNPs covered by query file: 14.499999999999998%\n",
      "- Found 1 (0.0022%) different reference variants. Adjusting...\n",
      "Inferring ancestry on query data...\n",
      "Phasing individual 1/1\n",
      "Writing phased SNPs to disk...\n",
      "Saving results...\n"
     ]
    }
   ],
   "source": [
    "mkdir -p output/snps_only\n",
    "uv run gnomix/gnomix.py output/EG100001.phased.hg19.vcf.gz output/snps_only chr22 True gnomix/pretrained_gnomix_models/chr22/model_chm_22.pkl > output/snps_only/gnomix.log 2>&1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b880ba78",
   "metadata": {},
   "source": [
    "Alternatively if you want to keep all variants, just normalize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6cb3e518",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/mismatch_removed/dup_removed/skipped:\t337580708/5583944/0/200294/0/0/0\n"
     ]
    }
   ],
   "source": [
    "bcftools norm --threads 32 -m -any -f reference/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta -o reference/EG100001.normalized.vcf.gz reference/EG100001.merged.g.vcf.gz\n",
    "\n",
    "bcftools index --threads 24 -t reference/EG100001.normalized.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "577391c6",
   "metadata": {},
   "source": [
    "Filter bam for only primary alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "29ab083f",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "samtools view -@ 64 -b -F 0x900 reference/EG100001.bam -o reference/EG100001.primary.bam\n",
    "\n",
    "samtools index -@ 64 reference/EG100001.primary.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e67b6d3e",
   "metadata": {},
   "source": [
    "Phase the normalized gVCF file and use only primary alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "28c085db",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "whatshap phase -o output/EG100001.phased.primaryonly.vcf.gz --reference=reference/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta reference/EG100001.normalized.vcf.gz reference/EG100001.primary.bam > WhatsHap.log 2>&1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d08262d2",
   "metadata": {},
   "source": [
    "Liftover phased VCF to hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ff1bb112",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to /tmp/bcftools.3mK7Zz\n",
      "INFO/AS_InbreedingCoeff is handled by AGR rule\n",
      "INFO/AS_QD is handled by AGR rule\n",
      "INFO/MLEAC is handled by AGR rule\n",
      "INFO/MLEAF is handled by AGR rule\n",
      "FORMAT/GP is handled by AGR rule\n",
      "FORMAT/PG is handled by AGR rule\n",
      "FORMAT/PL is handled by AGR rule\n",
      "Warning: input VCF includes symbolic alleles that might not properly lift over\n",
      "Lines   total/swapped/reference added/rejected:\t343586717/30993/7014/2756224\n",
      "Merging 19 temporary files\n",
      "Done\n",
      "Cleaning\n"
     ]
    }
   ],
   "source": [
    "docker run -it --rm -v \"$(pwd)\":/data -v /tmp:/tmp -w /data clarity001/bcftools-liftover:latest \\\n",
    "    /bin/bash -c \"bcftools +liftover -Ou output/EG100001.phased.primaryonly.vcf.gz -- -s reference/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta -f reference/hg19.fa -c reference/hg38ToHg19.over.chain.gz | bcftools sort -Oz -o output/EG100001.phased.primaryonly.hg19.vcf.gz -W=tbi\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a488c5c",
   "metadata": {},
   "source": [
    "Run the ancestry deconvolution tool (gnomix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "57e7dfee",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "...\n",
      "--------------------------------------------------------------------------------\n",
      "-----------------------------------  Gnomix  -----------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "When using this software, please cite: \n",
      "Helgi Hilmarsson, Arvind S Kumar, Richa Rastogi, Carlos D Bustamante, \n",
      "Daniel Mas Montserrat, Alexander G Ioannidis: \n",
      "\"High Resolution Ancestry Deconvolution for Next Generation Genomic Data\" \n",
      "https://www.biorxiv.org/content/10.1101/2021.09.19.460980v1\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "--------------------------------------------------------------------------------\n",
      "Launching in pre-trained mode...\n",
      "Loading model...\n",
      "Launching inference...\n",
      "Loading and processing query file...\n",
      "- Number of SNPs from model: 317408\n",
      "- Number of SNPs from file: 4652795\n",
      "- Number of intersecting SNPs: 85685\n",
      "- Percentage of model SNPs covered by query file: 27.0%\n",
      "- Found 9 (0.0105%) different reference variants. Adjusting...\n",
      "Inferring ancestry on query data...\n",
      "Phasing individual 1/1\n",
      "Writing phased SNPs to disk...\n",
      "Saving results...\n"
     ]
    }
   ],
   "source": [
    "mkdir -p output/normalized_primary_only\n",
    "uv run gnomix/gnomix.py output/EG100001.phased.primaryonly.hg19.vcf.gz output/normalized_primary_only chr22 True gnomix/pretrained_gnomix_models/chr22/model_chm_22.pkl > output/normalized_primary_only/gnomix.log 2>&1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
